### PR TITLE
Add logging into post and get requests

### PR DESCRIPTION
**Pull request by vitvakatu**
This change allows to debug POST/GET requests via `TestKitApi`. At the moment this functionality requires manual changes in specific projects' tests, which is kinda unfortunate.

Usage:
```
cd my-awesome-project-using-testkit
RUST_LOG=exonum_testkit=debug cargo test
```

Example output:
```
[1516698198 : 741] - [ DEBUG ] - exonum_testkit - GET request: http://localhost:3000/api/services/closed_service//v1/some/secret?count=1&from=1 Response: HTTP/1.1 200 OK
Content-Length: 132
Content-Type: application/json
```
```
[1516698198 : 743] - [ DEBUG ] - exonum_testkit - POST request: http://localhost:3000/api/services/secret_service//v1/secrets Body: {"body":{"content":{"content_hash":"0000000000000000000000000000000000000000000000000000000000000000","metadata":"metadata","user_id":"User"},"pub_key":"f4e5d9ecb97d7687080e17007758721c7941758ad7f292d1fd6f1c37bb91c5b1"},"message_id":2,"network_id":0,"protocol_version":0,"service_id":128,"signature":"cfd2d757ea39fddab369418fe73c364a858abc29790678b5c2a550efdc4e37225eec5c7b6eaba1ffb1a78e2ad7c4a0fac3ad2aa6f76016740756fafc216a2c09"} Response: "0a5194ac456bc4fadb83712d948ae67f0473ac34a469a5f1a7fc3f56e4bf131f"
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1224,6 +1226,8 @@ impl TestKitApi {
             resp.expect("Got unexpected `Err(..)` response")
         };
 
+        debug!("GET request: {} Response: {:?}", url, resp);
```
Comment (slowli) [inline src/lib.rs]: Nitpick: I'd use the original URL (possibly preceded by a slash) here and in POST request debugging to avoid questions like "where does `http://localhost:3000/` part come from".

It also could be nice to debug whether the request belongs to public or private API, I suppose; this probably requires adding an additional param to `get_internal` and `post_internal` (or replacing the `&Mount` param), say, an enum with variants `TestKitApiHandler::Private` and `TestKitApiHandler::Public`.

In summary, the output would look like

```none
Private GET /api/services/cryptocurrency/wallets
Response: [
  ...
]
```

**Review by slowli — COMMENTED** [review]
Concept LGTM as of a81580c1c2bb39627a5daa8ea91a0113ed4d6108. I think it could be possible to test this functionality by implementing a mock logger (or maybe there is already an existing one) and checking if it logs events as expected.

**Comment by slowli** [comment]
Another idea: maybe, it could make sense to implement logging as middleware wrapped around public / private handlers in `TestKitApi`? In this way, it would be possible to debug requests sent via `TestKitApi::public_mount` / `private_mount`. Also notice #23, which slightly changes `TestKitApi` internals; imo, it should be merged before this PR.

File: src/lib.rs (hunk 1 above)
Comment (stanislav-tkach) [inline src/lib.rs]: Some time ago we agreed not to use `debug` log level. Argumentation was the following: when one wants to debug some functionality one can add such logging and easily filter it in the whole project.

**Comment by matklad** [comment]
Could we use https://github.com/iron/logger?

I was in need of request logging when I was investigating hyper-layer performance. What I ended up doing was `RUST_LOG=hyper=debug cargo run --release `, which work OK for the real exonum, but doesn't work with testkit I guess.

File: Cargo.toml [hunk 2]
```diff
@@ -20,6 +20,7 @@ travis-ci = { repository = "exonum/exonum-testkit" }
 [dependencies]
 exonum = "0.4.0"
 futures = "0.1.14"
+log = "0.3"
```
Comment (alekseysidorov) [inline Cargo.toml]: Please update to 0.4

**Comment by vitvakatu** [comment]
Usage:
```
RUST_LOG=exonum_testkit=trace cargo test
```
Output:
```
[1516984712 : 252] - [ TRACE ] - exonum_testkit - GET api/services/secret//v1/payments/first_user?count=1&from=1
Response:
{
  "items": [
    {
      "metadata": "0",
      "total_amount": "0",
      "user_id": "first_user"
    }
  ],
  "total_count": 5
}

[1516984712 : 257] - [ TRACE ] - exonum_testkit - POST api/services/secret//v1/users
Body:
{"body":{"content":{"encrypted_secret_key":"8890d8f4b966e4d4ef50c8df071dbd1bab9a20636523ea4307a744646d21acb3a8b27b424b957f71ce6f2d48d94f60d1cc219ba1dff159844ded62aede4012b7","id":"User","metadata":"metadata","pub_key":"a8b27b424b957f71ce6f2d48d94f60d1cc219ba1dff159844ded62aede4012b7"},"pub_key":"344f3375011319f8f40a2b41f259185ad74d31aaa8785563e95cb9b96b046808"},"message_id":0,"network_id":0,"protocol_version":0,"service_id":128,"signature":"327463584167f49af45b9f50accd0402c2e47219499cdd6297488be0f12a23ec23f668b5710f482be66dff708d96ca9da4b045c88631bc755e71b7bd1c200d0b"}
Response:
"84ebe121f862730d3ceafa0154d0bc6efdfe8463e2600c1f5943e97818bceb79"
```

**Review by stanislav-tkach — DISMISSED** [review]

File: src/lib.rs [hunk 3]
```diff
@@ -1174,7 +1176,7 @@ impl TestKitApi {
         );
     }
 
-    fn get_internal<H, D>(handler: &H, url: &str, expect_error: bool) -> D
+    fn get_internal<H, D>(handler: &H, endpoint: &str, expect_error: bool, is_public: bool) -> D
```
Comment (alekseysidorov) [inline src/lib.rs:1179]: It just an internal method, but I suggest to use struct instead of boolean for better readability:
```rust
struct GetRequest {
    handler: H,
    endpoint: &str,
    expect_error: bool,
    is_public: bool
};
...
api.get_internal(GetRequest { handler, "endpoint", expect_error: true, is_public: false });
```

**Comment by slowli** [comment]
@vitvakatu Have you considered using the `logger` crate or other logging middleware? Are there any problems with this approach? (It seems that `logger` uses an older version of the `log` crate than it is used by Exonum; is this the problem?)

**Comment by vitvakatu** [comment]
@slowli there are several arguments against using `logger`:
- It's optional dependency
- In fact, it doesn't provide requested features (logging requests/responses), it's more useful for performance-related stuff.
- Different `log` versions may cause problems (mostly about filtering logs by modules)

So no, I'm not going to use it.

File: src/lib.rs (hunk 3 above)
Comment (vitvakatu) [reply src/lib.rs:1179]: I don't think it's strictly needed, but okay, let it be.

Comment (vitvakatu) [reply src/lib.rs:1179]: Oh well, actually... can we go without it? It just needs too many generics and lifetimes for these structs without much effort.

Comment (alekseysidorov) [reply src/lib.rs:1179]: Okay.

**Review by alekseysidorov — DISMISSED** [review]

**Comment by slowli** [comment]
@vitvakatu How about writing a dedicated middleware for the task then? My concern is that it's possible to call to the public / private handler without using `get*` or `post*` methods, so a middleware-based solution would be more robust. Subjectively, it would be prettier as well.

**Comment by vitvakatu** [comment]
@slowli I suggest to implement that functionality in another PR. In my opinion, the amount of users which will not use `post`/`get` methods is relatively small. At the same time, the basic logging infrastructure is vital and I really want to see it in the next release.

**Comment by slowli** [comment]
Got your point. Does the same reasoning apply to (lack of) tests as well? 🙃 I don't insist on tests if you feel that the PR should be merged ASAP, but it's not very nice to have untested functionality.

**Comment by vitvakatu** [comment]
After some consideration we've decided to cover tests in the follow-ups as well. I will also add info about this feature in Changelog and tutorial.

**Review by stanislav-tkach — DISMISSED** [review]

**Review by stanislav-tkach — APPROVED** [review]

**Review by alekseysidorov — APPROVED** [review]

**Review by Maria-Nosyk — APPROVED** [review]
